### PR TITLE
RIOT port for the MKW22D512 SiP and Phytec PBA-D-01 PhyWave Evaluations-Board

### PR DESCRIPTION
**Pull request by jfischer-phytec-iot**
This PR add support for Freescale MKW22D512 SiP and Phytec PBA-D-01 PhyWave Evaluations-Board.

### CODE REVIEW

**Comment by OlegHahm** [comment]
Do I see it correctly, that the license issue has been resolved?

@jfischer-phytec-iot, what's left on the Todo list?

**Comment by jfischer-phytec-iot** [comment]
@OlegHahm It is a CMSIS header with 3c-BSD license now.
Todo list? You mean #2188?
I have moved all peripheral driver in kinetis_common. I'm also porting RIOT to KW01Z128 and will make a PR soon. Internally, we already use kinetis_common for two MCUs (KW22D512 and KW01Z128). I think kinetis_common should include initially only the peripheral (uart, spi, i2c, hwtimer) drivers.
Currently I do not know what I can improve on that branch :-).

**Comment by OlegHahm** [comment]
I was mostly referring to the WIP label and what needs to be done in your opinion before this is ready.

**Comment by jfischer-phytec-iot** [comment]
We are waiting for/working on #2188. It's ready for the review as far as kinetis_common is finished.

**Comment by OlegHahm** [comment]
I see - shouldn't we change the `WIP` label to `Waiting for another PR` then?

**Comment by jfischer-phytec-iot** [comment]
@OlegHahm I've also thought of that :-). Would it be better to open own PR for kinetis_common and then change the label?

**Comment by OlegHahm** [comment]
Sounds sensible to me.

File: boards/pba-d-01-kw2x/include/periph_conf.h [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+ * @name Timer configuration
+ * @{
+ */
+#define TIMER_NUMOF                       (1U)
+#define TIMER_0_EN                        1
+#define TIMER_1_EN                        0
+#define TIMER_IRQ_PRIO                    1
+#define TIMER_DEV                         PIT
+#define TIMER_MAX_VALUE                   (0xffffffff)
+#define TIMER_CLOCK                       CLOCK_CORECLOCK
```
Comment (gebart) [inline boards/pba-d-01-kw2x/include/periph_conf.h]: Verify that this is the proper clock on the KW22. On the K60 the PIT is running on the Bus clock according to the K60 reference manual, chapter "5 Clock Distribution, 5.7 Module clocks"

Comment (jfischer-phytec-iot) [inline boards/pba-d-01-kw2x/include/periph_conf.h]: It is peripheral bus clock = MCGOUTCLK divided by 1 (OUTDIV2 in SIM_CLKDIV1) = coreclock.

**Comment by thomaseichinger** [comment]
As #2265 got merged, is this PR ready for review?

**Comment by jfischer-phytec-iot** [comment]
I want to test a few things and update it.
I will remove WIP when the time comes.

**Comment by thomaseichinger** [comment]
Sure, please ping me then.

**Comment by gebart** [comment]
You may need to update some periph_conf settings names since #2406
see https://github.com/gebart/RIOT/commit/26cd71f14a28e143c59bacda5145e7c43edd7299 and https://github.com/gebart/RIOT/commit/2e6578ba1e0a092700031e8d37217c8ee4d9cae9 for the changes necessary.

**Comment by OlegHahm** [comment]
I somehow managed to get my board in a non-working state. Flashing fails with the following warning:
```
START...
Warn : *********** ATTENTION! ATTENTION! ATTENTION! ATTENTION! **********
Warn : ****                                                          ****
Warn : **** Your Kinetis MCU is in secured state, which means that,  ****
Warn : **** with exception for very basic communication, JTAG/SWD    ****
Warn : **** interface will NOT work. In order to restore its         ****
Warn : **** functionality please issue 'kinetis mdm mass_erase'      ****
Warn : **** command, power cycle the MCU and restart OpenOCD.        ****
Warn : ****                                                          ****
Warn : *********** ATTENTION! ATTENTION! ATTENTION! ATTENTION! **********
Warn : Adapter returned success despite SSTICKYERR being set.
Error: CMSIS-DAP: Write Error (0x04)
in procedure 'reset' 
in procedure 'ocd_bouncer'

in procedure 'reset'
```
I tried `openocd -f ../../boards/pba-d-01-kw2x/dist/mkw22d512.cfg -c init -c "kinetis mdm mass_erase" -c reset -c exit` that apparently it didn't help. Any advice?

**Comment by gebart** [comment]
@OlegHahm Do you still have the image you flashed?

Start by examining the 16 bytes at offset 0x400 in that image (flash security sector)

See also http://cache.freescale.com/files/microcontrollers/doc/app_note/AN4507.pdf

**Comment by OlegHahm** [comment]
Thanks, apparently I managed to leave secured state with the help of this: http://bgamari.github.io/posts/2014-09-18-unbricking-a-secured-k20.html (pressing reset while firing `"kinetis mdm mass_erase"`). However, I still cannot properly flash the device. It hangs after
```
Info : Probing flash info for bank 0
Warn : flash configuration field erased, please reset the device
Warn : no flash bank found for address 1fff8000
wrote 22528 bytes from file /home/oleg/git/RIOT/examples/default/bin/pba-d-01-kw2x/default.elf in 12.146160s (1.811 KiB/s)
START...
Info : MDM: Chip is unsecured. Continuing.
```

**Comment by gebart** [comment]
You have some symbols whose load address is located in RAM it seems. 0x1fff8000 is SRAM_L. Examine your .map file and see what is going on there.

**Comment by OlegHahm** [comment]
Ah, I overlooked the warning. However, I at a first glance I'm not sure if anything's wrong here:
```
.ramcode        0x1fff8000        0x0 load address 0x0000515c
                0x1fff8000                . = ALIGN (0x4)
                0x0000515c                _ramcode_load = LOADADDR (.ramcode)
                0x1fff8000                _ramcode_start = .
 *(.ramcode*)
                0x1fff8000                . = ALIGN (0x4)
                0x1fff8000                _ramcode_end = .

.ARM.extab
 *(.ARM.extab* .gnu.linkonce.armextab.*)
                0x1fff8000                PROVIDE (__exidx_start, .)

.ARM.exidx
 *(.ARM.exidx* .gnu.linkonce.armexidx.*)
                0x1fff8000                PROVIDE (__exidx_end, .)

.eh_frame_hdr
 *(.eh_frame_hdr)

.eh_frame
 *(.eh_frame)

.gcc_except_table
 *(.gcc_except_table .gcc_except_table.*)
                0x1fff8000                . = ALIGN (0x4)
                0x1fff8000                _etext = .

.relocate       0x1fff8000       0xe4
                0x1fff8000                . = ALIGN (0x4)
                0x1fff8000                _data_load = LOADADDR (.relocate)
                0x1fff8000                _data_start = .
                0x1fff8000                _srelocate = .
 *(.ramfunc .ramfunc.*)
 *(.data .data.*)
 .data          0x1fff8000        0x0 /home/oleg/git/RIOT/examples/default/bin/pba-d-01-kw2x/kinetis_common/startup.o
 .data          0x1fff8000        0x0 /home/oleg/git/RIOT/examples/default/bin/pba-d-01-kw2x/kinetis_common/fault_handlers.o
 .data          0x1fff8000        0x0 /home/oleg/git/RIOT/examples/default/bin/pba-d-01-kw2x/cpu/interrupt-vector.o
 .data          0x1fff8000        0x0 /home/oleg/git/RIOT/examples/default/bin/pba-d-01-kw2x/cpu/syscalls.o
 .data.heap     0x1fff8000        0x4 /home/oleg/git/RIOT/examples/default/bin/pba-d-01-kw2x/cpu/syscalls.o
```

The rest after this address is `.data` and `.bss`, so I guess that's fine.

**Comment by jfischer-phytec-iot** [comment]
@OlegHahm 
As the Joakim wrote, you can lock the Kinetis (offset 0x400 with 16 inappropriate values :smile:). Behave like it shows, if the WDT is not disabled. It resets regularly the CPU and confuses the debuger.
Please try following:

1.
- Hold the reset button and execute ```make flash``` in a project directory
- Release reset button just after second "Info : MDM: Chip is unsecured. Continuing."

2.
- start openocd (```openocd -f ../../boards/pba-d-01-kw2x/dist/mkw22d512.cfg```)
- start telnet (```telnet localhost 4444```)
- (telnet) ```reset halt```
- (telnet) ```flash write_image /home/path/project/demo.elf```
- (telnet) ```reset run```

**Comment by jfischer-phytec-iot** [comment]
@thomaseichinger ping :checkered_flag:

**Comment by jfischer-phytec-iot** [comment]
Maybe someone has time for the Review ? :smile:

File: boards/pba-d-01-kw2x/include/board.h [hunk 2]
```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2014 Freie Universität Berlin
```
Comment (kaspar030) [inline boards/pba-d-01-kw2x/include/board.h]: You're working for FU now? :)

**Comment by kaspar030** [comment]
Looks OK. And I guess noone but Phytec has such a Phytec board, can we just merge this?

**Comment by OlegHahm** [comment]
No, if I'm not confusing the boards, we have some in Berlin, Hamburg and I have one with me in Paris. Last time I tested (months ago) it worked like charm (apart from the flashing issue). Can check tomorrow again.

**Comment by PeterKietzmann** [comment]
Yes we've got one of the eval boards. I used this PR several times when evaluation the sensor drivers. Never experienced any problems. Even if I can't ACK or test in detail atm I would never NACK this PR :-)

**Comment by jfischer-phytec-iot** [comment]
@kaspar030 Contact me if you want to have the board.

**Comment by OlegHahm** [comment]
For some reason the RTC seems to be not working:
```
2015-05-12 10:41:03,836 - INFO # Welcome to RIOT!
rtc gettime
2015-05-12 10:41:05,915 - INFO # > rtc gettime
2015-05-12 10:41:05,915 - INFO # 
2015-05-12 10:41:05,916 - INFO # 1970-01-01 00:00:00
rtc init
2015-05-12 10:41:09,162 - INFO # > rtc init
2015-05-12 10:41:09,163 - INFO # 
rtc poweron
2015-05-12 10:41:12,186 - INFO # > rtc poweron
2015-05-12 10:41:12,187 - INFO # 
rtc gettime
2015-05-12 10:41:14,443 - INFO # > rtc gettime
2015-05-12 10:41:14,443 - INFO # 
2015-05-12 10:41:14,444 - INFO # 1970-01-01 00:00:00
rtc gettime
2015-05-12 10:41:17,979 - INFO # > rtc gettime
2015-05-12 10:41:17,979 - INFO # 
2015-05-12 10:41:17,980 - INFO # 1970-01-01 00:00:00
```

**Comment by OlegHahm** [comment]
Apart from the RTC, everything's (still) fine. I think we are ready to merge, once the RTC issue is solved.

**Comment by jfischer-phytec-iot** [comment]
@OlegHahm I think you have a (alpha) module on the board whitout RTC-Crystal :-(
If it is populated, RTC should work.

**Comment by jremmert-phytec-iot** [comment]
@OlegHahm The new boards (those from the Embedded World) should have a RTC-Crystal. I just tested the RTC:

```
rtc init
2015-05-12 17:23:58,666 - INFO # > rtc init
2015-05-12 17:23:58,667 - INFO # 
rtc gettime
2015-05-12 17:24:05,708 - INFO # > rtc gettime
2015-05-12 17:24:05,709 - INFO # 
2015-05-12 17:24:05,710 - INFO # 1970-01-01 00:00:07
rtc gettime
2015-05-12 17:24:15,485 - INFO # > rtc gettime
2015-05-12 17:24:15,486 - INFO # 
2015-05-12 17:24:15,486 - INFO # 1970-01-01 00:00:16
rtc gettime
2015-05-12 17:24:19,091 - INFO # > rtc gettime
2015-05-12 17:24:19,092 - INFO # 
2015-05-12 17:24:19,093 - INFO # 1970-01-01 00:00:20
```

**Comment by jfischer-phytec-iot** [comment]
Rebased on master and added some last minute fixes.

**Comment by OlegHahm** [comment]
You squashed the last minute changes already?

**Comment by jfischer-phytec-iot** [comment]
No

**Comment by OlegHahm** [comment]
Which commits include these fixes?

**Comment by jfischer-phytec-iot** [comment]
cpu/kinetis_common/ldscripts/kinetis-noramcode.ld: adjust stack size 	67c9899
add auto init module 	fccbb18
boards/pba-d-01-kw2x: set default CPU_MODEL = kw21d256 	e463ba6
cpu/kw2x: add kw21d256 and kw21d512 cpu models 	daab6e6

File: cpu/kinetis_common/ldscripts/kinetis-noramcode.ld [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+__sram_length = LENGTH(sram);
+__sram_end    = __sram_start + __sram_length;
+_ramcode_start = 0x0;
+_ramcode_end = 0x0;
+_ramcode_load = 0x0;
+
+/* Define the default stack size for interrupt mode. As no context is
+   saved on this stack and ISRs are supposed to be short, it can be fairly
+   small. 512 byte should be a safe assumption here */
+STACK_SIZE = DEFINED(STACK_SIZE) ? STACK_SIZE : 0x800;
```
Comment (OlegHahm) [inline cpu/kinetis_common/ldscripts/kinetis-noramcode.ld]: Just curious: 2k of stack size for the ISRs? Being very conservative here and optimize later or are these really necessary?

**Comment by OlegHahm** [comment]
Apart from the open question and squashing: ACK!

File: cpu/kinetis_common/ldscripts/kinetis-noramcode.ld (hunk 3 above)
Comment (jfischer-phytec-iot) [inline cpu/kinetis_common/ldscripts/kinetis-noramcode.ld]: oh yes, we really need so much. If I find time, I will to examine it more closely.

**Comment by jremmert-phytec-iot** [comment]
I had (and still have) a similar issue in combination with the SAMR21-xpro board. The default stacksize there is 0x200, which leads to crashes when I test the ```driver_at86rf2xx``` application (0x400 works there, including printfs in kernel_init). Maybe these issues are related to our toolchain and library settings.

**Comment by OlegHahm** [comment]
Wasn't there even a ticket for this?

**Comment by OlegHahm** [comment]
For this PR, please go on and squash and merge at will.

File: cpu/kinetis_common/ldscripts/kinetis-noramcode.ld (hunk 3 above)
Comment (jremmert-phytec-iot) [inline cpu/kinetis_common/ldscripts/kinetis-noramcode.ld]: Previously, I should have posted the comment below here, sorry. Haven´t found an open ticket for this issue. But the issue on samr-21 only occured to me, so we should first investigate the toolchain before opening a new issue.

**Comment by jfischer-phytec-iot** [comment]
fccbb18 makes travis unhappy (failed: driver_at86rf2xx, driver_xbee, unittests)
I will remove it and squash. After #2756 has merged, we can open a PR for fccbb18.

**Comment by OlegHahm** [comment]
agreed

**Comment by jfischer-phytec-iot** [comment]
travis was canceled :confused: ?

**Comment by OlegHahm** [comment]
I killed some of the outdated builds in Travis - if I accidentally killed the most current one, too, I'm sorry, but apparently it's running again.

**Comment by jremmert-phytec-iot** [comment]
As https://www.traviscistatus.com/ states, there was an issue today. Maybe it was related to that somehow.

**Comment by jremmert-phytec-iot** [comment]
Travis failed because the testapp for the ```isl29125``` driver uses ```GPIO5```, which is per default disabled in our board.

**Comment by OlegHahm** [comment]
How about either modifying the Makefile with a special case for you board or (what I would prefer) blacklist the application?

**Comment by jremmert-phytec-iot** [comment]
I think we might not even need a special case for our board. Since GPIO_5 is just the default value we could change it to GPIO_0? This would also be consistent with the ng_at86rf2xx driver that counts the defaults from GPIO_0. GPIO_0 is defined by default and should be defined on all boards. As the comment in the Makefile (tests/driver_isl29125) says, this seems to be just a random value "*set random default*".

**Comment by OlegHahm** [comment]
Sounds good, too. @LudwigOrtmann, it's "your" driver. Any reason against changing this?

**Comment by jfischer-phytec-iot** [comment]
I have reorganized the GPIOs, again, a few more can not hurt.

**Comment by jremmert-phytec-iot** [comment]
Oh, the changes in ```periph_conf.h``` that were neccessary to build the kw2xrf app are included in the postponed fccbb18 commit. However, if you comment-in GPIO 24 conf everything should be fine.

```
/* GPIO channel 24 config */
/*
#define GPIO_24_DEV         KW2XDRF_GPIO
#define GPIO_24_PORT        KW2XDRF_PORT
#define GPIO_24_PORT_BASE   KW2XDRF_PORT_BASE
#define GPIO_24_PIN         KW2XDRF_PCS0_PIN
#define GPIO_24_CLKEN()     KW2XDRF_PORT_CLKEN()
#define GPIO_24_IRQ         KW2XDRF_PORT_IRQn
#define KW2XRF_CS_GPIO      GPIO_24
*/
```

**Comment by jfischer-phytec-iot** [comment]
my fault, i will update 470781b.

**Comment by jfischer-phytec-iot** [comment]
@jremmert-phytec-iot 
https://github.com/jremmert-phytec-iot/RIOT/commit/3287e71d45da591bc54f4400700004287014002d#diff-26037c789932d7c8d7bea2a82565588dR8
I do not want to enable GPIO_24 because we are using hardware CS per default (KW2XRF_SHARED_SPI=1).

**Comment by jremmert-phytec-iot** [comment]
We have (KW2XRF_SHARED_SPI=0) as default. But yes, that means hardware CS. 

I have also thought about that, but didn´t saw an alternative. How should we handle the parameter ```KWRF_CS``` then? We whould have to introduce two different parameter set of kw2xrf_init; i) without cs_pin and ii) with cs_pin and distinguish and let the preprocessor decide what function to use. I wanted to avoid that. 
```
 54 int kw2xrf_spi_init(spi_t spi, spi_speed_t spi_speed,
 55                     gpio_t cs_pin)
```

Do you have other solutions in mind?

**Comment by LudwigOrtmann** [comment]
Just define a specific GPIO for your board in the drivers Makefile.

**Comment by jfischer-phytec-iot** [comment]
@OlegHahm ack & squash ?

File: boards/pba-d-01-kw2x/dist/flash.sh [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+
+# flash without verify
+#if [ $x = "y" ] || [ $x = "v" ];# -o $x -eq v ];
+#then
+echo -e "${red}Flashing device, do not disconnect or reset the target until this script exits!!!${NC}"
+openocd -f "$RIOTBASE/boards/pba-d-01-kw2x/dist/mkw22d512.cfg" \
+	-c "init" \
+	-c "reset halt" \
+        -c "flash write_image erase $elffile" \
```
Comment (OlegHahm) [inline boards/pba-d-01-kw2x/dist/flash.sh]: superfluous space

**Comment by OlegHahm** [comment]
Yes, please fix the tiny indentation thing and squash. ACK

**Comment by jfischer-phytec-iot** [comment]
GO?

**Comment by OlegHahm** [comment]
Push the button!

**Pull request merged**